### ISSUE TITLE
Task adaptation from Cosmos3-Edge-Policy-DROID: is warm-starting with the trained action heads supported, and what compute should we expect?

### ISSUE THREAD

**Issue by erykpawelek** [NVIDIA/cosmos-framework#214]
### Context

We are an academic team at AGH University of Science and Technology (Kraków, Poland), using the ACK Cyfronet HPC centre. We want to adapt the released Edge policy to **one additional manipulation task** on the **same embodiment** it was trained for.

### What we have already read

So the answers below are not in the docs we have seen:

- `docs/action_policy_droid_posttrain.md`
- The "Post-Train NVIDIA Cosmos 3 Edge for On-Device Robot Control" blog post
- The `nvidia/Cosmos3-Edge` and `nvidia/Cosmos3-Edge-Policy-DROID` model cards
- The Cosmos Policy paper (arXiv 2601.16163)

We understand that the published recipe initialises from the **Cosmos3-Edge base checkpoint** with freshly initialised action heads (`keys_to_skip_loading = ["action2llm", "llm2action", "action_modality_embed"]`), and that the validated run is 64 nodes × 4× GB200 (~17.4K GB200-hours). We also understand the blog states that for a DROID-like Franka setup "the primary change is the dataset path".

Our question is about the case the docs do not cover: starting from the **already post-trained policy** rather than the base checkpoint.

### Our setup

| | |
|---|---|
| Embodiment | Franka Panda — same as DROID. `action_space=joint_pos`, 8-D |
| Task | One new task: rotating a quarter-turn ball-valve lever ~90° |
| Data | <700 episodes, ~15 s each at 15 fps ≈ 160k frames, LeRobotDataset v3.0, 3 cameras |
| Cluster A | Cyfronet Athena — 8 × A100-SXM4-40GB per node, x86_64, there are small job time limits |
| Cluster B | Cyfronet Helios — 4 × GH200 96 GB per node, **aarch64**, there are small job time limits |

### Questions

**1. Warm start.** Is there a supported way to initialise training from `nvidia/Cosmos3-Edge-Policy-DROID` while **keeping** the trained action heads — i.e. not passing `keys_to_skip_loading` for `action2llm` / `llm2action` / `action_modality_embed`? Is this a configuration change only, or does it require code changes?

**2. Expected compute for task adaptation.** If warm-starting is supported, for a single new task on an already-supported embodiment with ~700 episodes, what iteration count and what wall-clock or GPU-hours should we expect? **An order of magnitude would already be very useful.** We understand the 17.4K GB200-hour figure describes reproducing the release policy from the base checkpoint on the full 76k-trajectory DROID dataset, and we assume task adaptation is far smaller — but we have no published reference point at all.

**3. Minimum viable hardware.** Does full-parameter action SFT of Cosmos3-Edge fit on **one node of 8 × A100 40 GB** with FSDP full sharding and `activation_checkpointing.mode = "full"`? If not, what is the smallest configuration you have validated?

**4. Parameter-efficient modes.** Is `[model].lora_enabled` supported for the **action pathway**, or is LoRA VFM-only in the shipped recipes? Separately, is restricting trainable parameters to the action heads via `[optimizer].keys_to_select` a validated mode for task adaptation, and does it converge acceptably?

**5. Demonstration count.** Do you have internal guidance on the minimum number of demonstrations for a new task on an already-supported embodiment? The Cosmos Policy paper reports 50 demos per task on RoboCasa — does that data efficiency carry over to Cosmos3-Edge-Policy-DROID?

**6. aarch64.** Is `cosmos-framework` training supported on GH200 (Grace ARM CPU + Hopper GPU)? Any known issues, or wheels we should use?

Happy to share results publicly once we have them. Thank you.

**Comment by lfengad** [comment]
Thanks for the detailed question and for clearly separating full DROID post-training from single-task adaptation.

The short answer is that the current code has the underlying checkpoint-loading support needed to retain trained action weights, but the OSS release does not yet provide or validate an end-to-end Cosmos3-Edge-Policy-DROID task-adaptation recipe. In particular, this is not currently a TOML-only change with the shipped artifacts.

1. **Warm start**

   After converting a compatible policy checkpoint to DCP, the loader can preserve its action weights: the action-related patterns (`action2llm`, `llm2action`, `action_modality_embed`, and `action_pos_embed`) must not be present in `checkpoint.keys_to_skip_loading`. `net_ema.` may remain skipped; the checkpoint implementation can initialize EMA from the loaded regular network weights.

   However, the shipped `action_policy_droid_nano` experiment constructs the model from `NANO_MODEL_CONFIG`. Although the repository also contains `EDGE_MODEL_CONFIG`, it is not currently wired into an `action_policy_droid_edge` experiment. Therefore the existing Nano recipe must not be pointed at an Edge checkpoint unchanged. Supporting Edge task adaptation requires an Edge-specific experiment/configuration in code, followed by validation.

   The current wording in `docs/action_policy_droid_posttrain.md` is ambiguous here and should be clarified: the shipped DROID reproduction experiment is Nano-specific, even though shared checkpoint/model infrastructure also contains Edge support.

2. **Expected compute**

   We do not currently have a validated Cosmos3-Edge-Policy-DROID single-task adaptation benchmark, so we cannot provide a reliable iteration count, wall-clock time, or GPU-hour estimate. The 10,000-iteration, 256-GPU reference is for full DROID post-training and should not be extrapolated directly to a warm-started single-task run.

   The nearest public action-policy reference is LIBERO: the published recipes use 2,000 iterations for LIBERO-10 (peaking around 1,500) and 5,000 for the four-suite mixture (around 4,500 to converge). Those numbers are only scale references; the dataset, action representation, model tier, batch size, and evaluation environment differ substantially.

   For bring-up, we recommend saving and running closed-loop evaluation frequently rather than selecting a stopping point from training loss alone. Once step time has been measured on the target system, the mechanical estimate is:

   ```text
   GPU-hours = step_time_seconds * optimizer_steps * GPU_count / 3600
   ```

3. **Minimum viable hardware**

   Ampere/BF16, full-shard FSDP, full activation checkpointing, and per-rank sample caps are supported by the framework. We have not, however, validated full-parameter Cosmos3-Edge action SFT on 8 x A100 40 GB, so we cannot claim that the reference micro-batch fits.

   The 480p three-camera action sequence is activation-heavy. A reasonable memory bring-up is shard degree 8, `activation_checkpointing.mode = "full"`, and a small `dataloader_train.max_samples_per_batch`, increasing it only after measuring peak memory. The smallest documented Edge generator-SFT validation is 4 x GB200, but that is not an action-policy validation.

4. **Parameter-efficient modes**

   `model.lora_enabled` is implemented for the VFM generation transformer. The default targets are the generation-pathway Q/K/V/O projections; no LoRA action-policy recipe is currently shipped or validated.

   More specifically, `action2llm` and `llm2action` are embedding-backed `DomainAwareLinear` modules, while the current LoRA injector wraps `nn.Linear`. Enabling the existing LoRA path therefore does not add adapters directly to the action heads.

   Restricting `optimizer.keys_to_select` to the action-head parameter names is mechanically supported by the optimizer filter, but its DROID closed-loop convergence has not been validated. The released DROID recipe trains the action heads together with the generation transformer, timestep embedding, and vision input/output projections, so head-only training should currently be considered experimental.

5. **Demonstration count**

   We do not have a validated minimum demonstration count for DROID task adaptation. The 50-demonstration RoboCasa result should not be assumed to transfer directly: initial-state coverage, visual diversity, action noise, and task-success labeling can matter more than the raw episode count.

   Approximately 700 episodes are sufficient for a meaningful experiment, but we cannot predict the resulting closed-loop success rate. We recommend holding out initial conditions and object/camera variations rather than evaluating only on randomly held-out frames from the same episodes.

6. **aarch64 / GH200**

   Linux aarch64 is supported by the dependency configuration, and GH200 is within the supported Hopper generation. The recommended route is the NGC PyTorch base image followed by the standard `cu130-train` installation. We have not separately validated this exact Edge action-policy adaptation on GH200, but there is no intended x86-only restriction in the training framework.

   For short scheduler limits, DCP checkpoint/resume is supported: relaunch against the same output root/job identity to continue from the latest saved checkpoint.

We would be very interested in any results you can share. A validated Edge action-policy recipe and measured single-node memory/throughput numbers would also be valuable additions to the documentation.
